### PR TITLE
Follow conversation by push notification.

### LINKED ISSUE #15460: Follow conversation by push notification.
Currently the reader offers the possibility to follow by email a conversation happening as comments to a post. In addition to that we want to allow a user to be notified of a new comment on a post by push notification.

Here below the designs for iOS (I will add the Android screens later).

image

Note that the image reports the threaded comments with the new design but this issue is going to be tackled before the new threaded comments is in place.

### PR DESCRIPTION
**Pull request by develric**
Fixes #15460

References: `pctCYC-9x-p2`

This PR adds to the reader the possibility to activate push notifications related to comments after following a post comments conversation by email already.

https://user-images.githubusercontent.com/47797566/137237202-919e2e15-d218-47f9-b36e-21db5408ffc0.mov

Note 1: the feature is currently behind a feature flag and you should enable it to test. [UPDATE: the feature flag has been activated in #15469
Note 2: the image in the linked issue reports the threaded comments with the new design but this issue is being tackled before the new threaded comments is in place.

Dedicated unit testing will be added in a separate PR. [UPDATE: added [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15469)]

### New feature testing
- Go to the reader and select one of the tabs (Following/Discover etc...)
- Open the comments on a post from the reader stream (do the same later from the comment details)
- Check the option menu shows a `FOLLOW` shimmer item until the following status is recovered. 
- If you are not following the post comments already, select the FOLLOW menu and check a snackbar notifies you of the action result
- If it succeeded the snackbar present an action to enable push notifications; try to not activate push notifications and check you get only emails for new comments on that post
- try to activate the push notifications and check you get both emails and push notifications for new comments on that post
- Check you can enable and undo push notifications activation from the snackbar actions
- Check you receive notifications and email for new comments when push notifications are on, only emails when push notificaiton are off, no email nor push notifications when you unsubscribe from a post comments
- Check you get an error when in airplane mode (note that on errors on post following information rescue, the menu is placed in an unknown state as a dimmed FOLLOW button)
- Check a PTR on the comments list updates the state of the following menu
- Check you can change the push notification status from the bottom sheet switch and you can fully unfollow the conversation
- Check that if you are not logged in in .COM and you go to the reader, accessing the comments on a post does not present the follow menu

### Smoke testing
- With the feature flag enabled, smoke test the threaded comments; try to like and reply to a comment, post a new comment etc...
- Disable the feature flag and smoke test the previous follow by email functionality

NOTE: push notifications for replies to your comments are still managed as before, comments that are coming from push notifications specifically are identified with a `remark` wording

<p align=center>

</p>

## Regression Notes
1. Potential unintended areas of impact
N/A: the feature is behind a feature flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A: the feature is behind a feature flag.

3. What automated tests I added (or what prevented me from doing so)
N/A: the feature is behind a feature flag.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/res/layout/follow_action_layout.xml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        <!--<TextView
+            android:id="@+id/follow_button"
+            android:layout_width="wrap_content"
+            android:layout_height="match_parent"
+            android:layout_marginEnd="@dimen/margin_large"
+            android:layout_marginStart="@dimen/margin_large"
+            android:background="?android:attr/actionBarItemBackground"
+            android:text="@string/reader_btn_follow"
+            android:textAppearance="?android:attr/actionMenuTextAppearance"
+            android:textColor="?android:attr/actionMenuTextColor" />-->
```
Comment (ashiagr) [inline WordPress/src/main/res/layout/follow_action_layout.xml]: 🔍 Minor: Can we remove this commented code?

File: WordPress/src/main/res/menu/threaded_comments_menu.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        android:id="@+id/follow_item"
+        android:title="@string/reader_btn_follow"
+        app:showAsAction="always"
+        app:actionLayout="@layout/follow_action_layout"/>
+    <item
+        android:id="@+id/manage_notifications_item"
+        android:title="@string/reader_btn_follow"
+        android:icon="@drawable/ic_notifications_white_24dp"
+        app:showAsAction="always"/>
+</menu>
```
Comment (ashiagr) [inline WordPress/src/main/res/menu/threaded_comments_menu.xml:14]: 🔍 Minor: This file can be formatted.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/utils/PostSubscribersApiCallsProvider.kt [hunk 3]
```diff
@@ -143,10 +182,14 @@ class PostSubscribersApiCallsProvider @Inject constructor(
         }
     }
 
-    private fun isFollowing(json: JSONObject?): PostSubscribersCallResult {
+    private fun getFollowingStateResult(json: JSONObject?): PostSubscribersCallResult {
         return json?.let {
-            if (it.has("i_subscribe")) {
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/utils/PostSubscribersApiCallsProvider.kt:148]: 🔍  Minor: Can we extract `i_subscribe` and `receives_notifications` as constants considering they're used multiple times in the class?

File: WordPress/src/main/java/org/wordpress/android/ui/reader/utils/PostSubscribersApiCallsProvider.kt [hunk 4]
```diff
@@ -156,7 +199,7 @@ class PostSubscribersApiCallsProvider @Inject constructor(
     private fun canFollowComments(blogId: Long, json: JSONObject?): PostSubscribersCallResult {
         return json?.let {
             if (it.has("ID") && it.optLong("ID", -1) == blogId) {
-                Success(false)
+                Success(false, false)
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/utils/PostSubscribersApiCallsProvider.kt]: Suggestion: Now that we have two params for `Success` state, we might use [named argument](https://kotlinlang.org/docs/coding-conventions.html#named-arguments) syntax to improve readability.  Note that the AS IDE also provides this warning:
`Boolean literal argument without parameter name `

We can check all instances of `Success` where we're directly passing `true`/ `false` in this class and replace them with named args.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListViewModel.kt [hunk 5]
```diff
@@ -85,13 +144,31 @@ class ReaderCommentListViewModel
...
+                        ::enablePushNotificationsFromSnackbarAction
+                )
+            } else {
+                followCommentsHandler.handleFollowCommentsClicked(
+                        blogId,
+                        postId,
+                        askSubscribe,
+                        null
+                )
+            }
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListViewModel.kt]: Suggestion: It might be better to extract `onSuccessSnackbarAction` into a `var`, something like this:

```
val enablePushNotificationsFromSnackbarAction = followByPushNotificationFeatureConfig.isEnabled() && askSubscribe
followCommentsHandler.handleFollowCommentsClicked(
                    blogId,
                    postId,
                    askSubscribe,
                    if (enablePushNotificationsFromSnackbarAction) ::enablePushNotificationsFromSnackbarAction else null
            )
```

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java [hunk 6]
```diff
@@ -167,22 +177,56 @@ public void onCreate(Bundle savedInstanceState) {
...
-        mViewModel.getSnackbarEvents().observe(this, event ->
-                event.applyIfNotHandled(holder -> {
-                    WPSnackbar.make(mRecyclerView,
-                            mUiHelpers.getTextOfUiString(this, holder.getMessage()),
-                            Snackbar.LENGTH_LONG)
-                              .show();
-                    return Unit.INSTANCE;
-                })
-        );
+        mViewModel.getSnackbarEvents().observe(this, snackbarMessageHolderEvent -> {
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java]: 🔍  Minor:  Can we extract `ViewModel` observation logic into a separate method `initViewModel` considering `onCreate` method has grown in size (there's also an IDE warning for the method size)? 

PS: Same comment applies to `CommentNotificationsBottomSheetFragment.onViewCreated` where we can  separate out `ViewModel` and UI related logic.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderCommentsFollowUseCase.kt [hunk 7]
```diff
@@ -20,14 +22,16 @@ import org.wordpress.android.ui.utils.UiString
...
 class ReaderCommentsFollowUseCase @Inject constructor(
     private val networkUtilsWrapper: NetworkUtilsWrapper,
     private val postSubscribersApiCallsProvider: PostSubscribersApiCallsProvider,
     private val accountStore: AccountStore,
     private val readerTracker: ReaderTracker,
-    private val readerPostTableWrapper: ReaderPostTableWrapper
+    private val readerPostTableWrapper: ReaderPostTableWrapper,
+    private val followByPushNotificationFeatureConfig: FollowByPushNotificationFeatureConfig
 ) {
     private val FOLLOW_COMMENT_ACTION = "follow_action"
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderCommentsFollowUseCase.kt]: Suggestion: (Not related to changes in this PR) Can we convert these `vals` (`FOLLOW_COMMENT_ACTION`, `FOLLOW_COMMENT_ACTION_RESULT`, `FOLLOW_COMMENT_ACTION_ERROR`) to constants? There's also a `detekt` warning that will get fixed by this:

`Detekt - VariableNaming: Private variable names should match the pattern: (_)?[a-z][A-Za-z0-9]*`

File: WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderCommentsFollowUseCase.kt [hunk 8]
```diff
@@ -150,7 +230,9 @@ class ReaderCommentsFollowUseCase @Inject constructor(
 
     private enum class AnalyticsFollowCommentsAction(val action: String) {
         FOLLOW_COMMENTS("followed"),
-        UNFOLLOW_COMMENTS("unfollowed")
+        UNFOLLOW_COMMENTS("unfollowed"),
+        ENABLE_PUSH_NOTIFICATION("enable_push_notifications"),
+        DISABLE_PUSH_NOTIICATION("disable_push_notifications")
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderCommentsFollowUseCase.kt]: 🔍  Minor: There's a typo in `DISABLE_PUSH_NOTIICATION`.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java [hunk 9]
```diff
@@ -399,13 +445,100 @@ public void onEventMainThread(SuggestionEvents.SuggestionNameListUpdated event)
...
+                                });
+
+                                setMenuStatesByStateType(stateType,
+                                        followItem,
+                                        bellItem,
+                                        followText,
+                                        shimmerView,
+                                        isButtonEnabled,
+                                        uiState.isFollowing());
+                            }
```
Comment (ashiagr) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java]: Suggestion: There's a scope of moving some of this logic to VM from the Activity, including logic from `setMenuStatesByStateType` function (like creating a mapping between `FollowCommentsUiStateType` and states for `followItem` & `bellItem` ). Wdyt?

**Review by ashiagr — COMMENTED** [review]
Thanks for working on this task, I'm so very happy that we have this option to be notified in-app for the conversations we follow! 🎉 

I was able to receive notifications as per instructions, only noticed a couple of issues that I already discussed with you. As discussed, they might be linked to this known issue: p1633941032482100-slack-C01CW1VMLAF

1. I enabled in-app push notifications for conversation as mentioned in the instructions, but did not receive them in an instance when I had notifications disabled for the comments on my own (atomic) site under the `Notifications` tab. Once I enabled the setting, I started receiving them.

2. I disabled notifications for conversations, yet was able to receive them when someone commented on the thread. Comments needed moderation, once approved, new comments from the same user were displayed as remarked comments under the `Notifications` tab.    

    https://user-images.githubusercontent.com/1405144/137310743-dae7be00-d653-4ce7-972f-4287b4741d91.mp4

------------------------------------

Besides I left minor code-level comments, some are very easy to fix - detekt warnings etc or related to `Fragment/ Activity `to `ViewModel` logic refactoring.

Once tests are included for the feature, we can further test it exhaustively.

Hope this helps!

**Comment by develric** [comment]
Thanks so much for deep reviewing and testing this PR @ashiagr 🙇 . I'm going to work on the feedbacks and come back to you 👍 ! 

Hi @mattmiklic 👋 , I didn't ping you yesterday since I knew you were AFK and this is still behind feature flag, but happy if you could have a look design-wise to the current implementation while I cover the code feedbacks (the feature is behind a feature flag as said and you should be able to activate it installing the jalapeno APK from this CI. Just ping me if any issue with that 🙇 ). Thanks in advance sir 😄 !

File: WordPress/src/main/res/layout/follow_action_layout.xml (hunk 1 above)
Comment (develric) [reply WordPress/src/main/res/layout/follow_action_layout.xml]: 🤦‍♂️! ✅ 3a8933c7d32bf8b400a1c4493bbca5ea1af91588

File: WordPress/src/main/res/menu/threaded_comments_menu.xml (hunk 2 above)
Comment (develric) [reply WordPress/src/main/res/menu/threaded_comments_menu.xml:14]: bccbd56683a5a931b92dbdc68e0740d99895ff2d

File: WordPress/src/main/java/org/wordpress/android/ui/reader/utils/PostSubscribersApiCallsProvider.kt (hunk 3 above)
Comment (develric) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/utils/PostSubscribersApiCallsProvider.kt:148]: b66b4a6482335ce2658c37019ddbc720a5a6eae5

File: WordPress/src/main/java/org/wordpress/android/ui/reader/utils/PostSubscribersApiCallsProvider.kt (hunk 4 above)
Comment (develric) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/utils/PostSubscribersApiCallsProvider.kt]: ed5d7d650378c884beb14009507103aaf45041a0

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListViewModel.kt (hunk 5 above)
Comment (develric) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListViewModel.kt]: this is actually what I was doing initially 😄 , but for reasons I do not understood build was failing saying it was not able to find the ::enablePushNotificationsFromSnackbarAction function; strangely enough also the CI was failing the build for the same reason 🤷 ....I've just tried again and it worked locally...let's give it a try on the CI since of course that's cleaner 👍 ! fed5c60174fcb556fcf4925c1e4c22877ec7f08c

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java (hunk 6 above)
Comment (develric) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java]: ✅! I tried to not move too much things around to minimize issues. 7a2956fc2511dc0d9b8e4b20a926bf9a78b01a23 and 252413d4dd855cd742f3a7effb814ed24931922d

File: WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderCommentsFollowUseCase.kt (hunk 7 above)
Comment (develric) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderCommentsFollowUseCase.kt]: ✅ e6c661dcb43d2708df96e3289dd13882b335b9b8

File: WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderCommentsFollowUseCase.kt (hunk 8 above)
Comment (develric) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderCommentsFollowUseCase.kt]: ✅ 0f5b8d67d861ea2233f794c434a8b85f7204a13f

**Comment by mattmiklic** [comment]
Looks like the design was implemented perfectly! 👨‍🍳👌 Nice work with this one.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java (hunk 9 above)
Comment (develric) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java]: ✅  cd5e35d3f715d67025b0ceeaa0bed4994455d3f4

**Comment by develric** [comment]
Hey @ashiagr 👋 , I followed up to your feedbacks, thanks again 🙇 . This should be ready for another pass now.

Re the 2 issues you reported, I was not able to reproduce on my side actually. If possible we can maybe sync check when we are both online in case I'm missing any step to reproduce 🙇 

Note: I'm keeping this PR branch like a feature branch creating a couple PRs against it. Let's align on the 2 mentioned issue before merge to develop, thanks.

**Review by ashiagr — APPROVED** [review]
Thanks for all the updates, @develric! All looks good to me.  👍 

As discussed in our call, let's create separate issues to test those edge cases later on as they're not consistently reproducable. 

I'm approving the PR, please feel free to merge once unit tests are integrated.
